### PR TITLE
🚸 Add PupylCommandLineInterface

### LINKED ISSUE #35: ✨ Command line interface
* Create a command line interface for the library, so everyone can issue operations from the shell;
* It must expose the main features.

### PR DESCRIPTION
**Pull request by 92amartins**
This is a first attempt to close #35 and provide a friendlier UI for the users.

## Usage
```sh
python pupyl.py --data_dir samples/images.tar.xz
```

Some considerations for discussion:

- [x] pupyl.py in the root folder is not something that I particularly like, but seems necessary to provide the interface: `python pupyl --data_dir my_image_dir --serve`;
- [x] no tests yet. Maybe after some discussion on the interface testing strategy might get clearer;
- [x] support `pupyl [OPTIONS] [ETC]` instead of `python pupyl.py [OPTIONS] [ETC]`;
- [x] I guess we should update README 🤔 ;
- [x] Deal with script path resolution (the `CLI` script should be called everywhere);
- [x] Make `data_dir` `CLI` parameter not mandatory, creating the `pupyl` assets on the directory (instead to the underlying temporary directory) on the case of this parameter was omitted (with a disclaimer);
- [x] Resolve script to the call `pupyl [OPTIONS]`, not `pupyl-cli [OPTIONS]`.

@policratus, I think we can greatly improve on that, would love to hear your thoughts.

### CODE REVIEW

File: pupyl.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+"""
```
Comment (policratus) [inline pupyl.py:1]: @92amartins, as a standalone application, maybe it could be better executing as an independent script, by adding:

```suggestion
#!/usr/bin/env python3
"""
```
as described in [Should I put #! (shebang) in Python scripts, and what form should it take?](https://stackoverflow.com/questions/6908143/should-i-put-shebang-in-python-scripts-and-what-form-should-it-take).

But maybe we should check if this shebang is OS agnostic (for example, it'll work on Windows?).

By doing this, we can rename the file from `pupyl.py` to just `pupyl`, and can call like this:

```shell
pupyl [OPTIONS] [ETC]
```
instead of
```shell
python pupyl.py [OPTIONS] [ETC]
```

File: pupyl.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+
+if __name__ == "__main__":
+    cli = PupylCommandLineInterface()
+    args = cli.parser.parse_args()
+    pupyl = PupylImageSearch()
+
+    pupyl.index(
+        args.data_dir
+    )
```
Comment (policratus) [inline pupyl.py:17]: @92amartins , as it's constants, prefer names in uppercase (more: [constants](https://www.python.org/dev/peps/pep-0008/#constants)):

```suggestion
    CLI = PupylCommandLineInterface()
    ARGS = CLI.parser.parse_args()
    PUPYL = PupylImageSearch()

    PUPYL.index(
        ARGS.data_dir
    )
```

File: pupyl/cli.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class PupylCommandLineInterface:
+    def __init__(self):
+        self.parser = argparse.ArgumentParser(
+            description='Manage pupyl cli arguments.'
+        )
+        self.parser.add_argument(
+            '--data_dir', help='data directory for image files'
+        )
```
Comment (policratus) [inline pupyl/cli.py:13]: @92amartins, the argument `data_dir` should be positional, not optional. Maybe it's better:

```suggestion
        self.parser.add_argument(
            'data_dir', help='Data directory for image files.'
        )
```

There's an idea to let `data_dir` as an optional argument, but we should clearly advise (as in [README.md](https://github.com/policratus/pupyl#-usage)) the user that in this case (of not filling this argument) all assets will be created on their temporary directory (and probably will be lost on next reboot).

So, for instance:

```shell
# Creates all assets on the system temporary directory
pupyl --index /path/to/images
```
or
```shell
# Creates all asset on /path/to/assets
pupyl --index /path/to/images --data_dir /path/to/assets
```

**Review by policratus — COMMENTED** [review]
In general, we should discuss how the overall `CLI` should work. What you think of something like the following?

```shell
pupyl index /path/to/images
pupyl index /path/to/images --data_dir /path/to/assets
pupyl serve --data_dir /path/to/assets
pupyl search /path/to/a/query/image
```

File: pupyl/cli.py (hunk 3 above)
Comment (92amartins) [reply pupyl/cli.py:13]: For the sake of simplicity, I think we should go for mandatory, but support `data_dir` as a optional argument in the future.

**Comment by 92amartins** [comment]
> In general, we should discuss how the overall `CLI` should work. What you think of something like the following?
> 
> ```shell
> pupyl index /path/to/images
> pupyl index /path/to/images --data_dir /path/to/assets
> pupyl serve --data_dir /path/to/assets
> pupyl search /path/to/a/query/image
> ```

In this PR, I'm assuming a single command to "win them all".
I like the idea of having these specific commands, but I don't know how we could implement that 🤔  Any pointers on that?

File: pupyl/cli.py (hunk 3 above)
Comment (92amartins) [reply pupyl/cli.py:13]: Regarding the suggestion on the help text, I just followed the `-h` defaults (no uppercase, no end stop).
I think we should stick to it.

positional arguments:
  data_dir    data directory for image files

optional arguments:
  -h, --help  show this help message and exit

Comment (92amartins) [reply pupyl/cli.py:13]: Lastly, could you clarify on the difference between `/path/to/images` and `/path/to/assets`? My understanding is that the first refers to _input images_ and the second is some kind of intermediate files for pupyl's internal representation.

File: pupyl.py (hunk 1 above)
Comment (92amartins) [reply pupyl.py:1]: Tried to follow this recommendation and couldn't make that work on macOS. I use pyenv so maybe there's a trouble with shims. Can you make that work in linux?

Comment (policratus) [reply pupyl.py:1]: Yep, maybe it's something to do with `pyenv`, because I tested on Linux without issues. As Windows systems doesn't support shebangs, we should not worry about this system right now.

If you agree, we can focus to support Linux and MacOS for a while, trying to understand what's happening with `pyenv` on MacOS (unfortunately I don't have one right at hand).

File: pupyl/cli.py (hunk 3 above)
Comment (policratus) [reply pupyl/cli.py]: > For the sake of simplicity, I think we should go for mandatory, but support `data_dir` as a optional argument in the future.

Totally agree! 🚀

Comment (policratus) [reply pupyl/cli.py]: > Regarding the suggestion on the help text, I just followed the `-h` defaults (no uppercase, no end stop).
> I think we should stick to it.
> 
> ```
> positional arguments:
>   data_dir    data directory for image files
> 
> optional arguments:
>   -h, --help  show this help message and exit
> ```

You're right, my bad!

Comment (policratus) [reply pupyl/cli.py]: > Lastly, could you clarify on the difference between `/path/to/images` and `/path/to/assets`? My understanding is that the first refers to _input images_ and the second is some kind of intermediate files for pupyl's internal representation.

Yep, kinda.

**TL;DR**
------------
**`/path/to/images` is like a directory with CSVs to use as input data for a database; `/path/to/assets` is where the database internal files are stored.**

**TS:WM**
------------
The path reference used on [pupyl.search.PupylImageSearch.index()](https://github.com/policratus/pupyl/blob/c79f68d2d1ab0651e0a0316da963da8687315cc2/pupyl/search.py#L105) (and other subsequent methods) describes a _source_, or in other words the _input data_ or the images that must be inserted and indexed. Hence, references to the example path `/path/to/images` are exactly this, image that will be read.

References to `data_dir` (as in [pupyl.search.PupylImageSearch.__init__(self, data_dir=None, **kwargs)](https://github.com/policratus/pupyl/blob/c79f68d2d1ab0651e0a0316da963da8687315cc2/pupyl/search.py#L29l)) are related to internal representations (and files) of a created database, which contains copies of `/path/to/images` images, index files, metadata, etc.

**Comment by policratus** [comment]
> > In general, we should discuss how the overall `CLI` should work. What you think of something like the following?
> > ```shell
> > pupyl index /path/to/images
> > pupyl index /path/to/images --data_dir /path/to/assets
> > pupyl serve --data_dir /path/to/assets
> > pupyl search /path/to/a/query/image
> > ```
> 
> In this PR, I'm assuming a single command to "win them all".
> I like the idea of having these specific commands, but I don't know how we could implement that  Any pointers on that?
> 
> UPDATE:
> 
> Did some research. I think [subcommands](https://docs.python.org/3/library/argparse.html#sub-commands) will do the trick.

Awesome, makes all the sense! 🥇

File: pupyl/cli.py (hunk 3 above)
Comment (92amartins) [reply pupyl/cli.py]: Nice. Thanks for the clarifications o/

File: pupyl/cli.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+import argparse
+
+
+class PupylCommandLineInterface:
+    def __init__(self):
+        self.parser = argparse.ArgumentParser(
+            description='Manage pupyl cli arguments.'
+        )
+        self.subparsers = self.parser.add_subparsers(dest='command_name')
```
Comment (policratus) [inline pupyl/cli.py:11]: @92amartins, running the test suites, seems like there's an issue on subparsers:

```python
usage: py.test [-h] {index,serve} ...
py.test: error: cannot have multiple subparser arguments
```

File: pupyl.py (hunk 1 above)
Comment (policratus) [reply pupyl.py:1]: What you think @92amartins? Should we focus on Linux on this case?

File: tests/test_cli.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
+"""Unit tests for the command line interface."""
+
+from pupyl.cli import PupylCommandLineInterface
+
+CLI = PupylCommandLineInterface()
```
Comment (policratus) [inline tests/test_cli.py:5]:
```suggestion
from pupyl.cli import PupylCommandLineInterface


CLI = PupylCommandLineInterface()
```

**Review by policratus — COMMENTED** [review]
:rocket: We're getting there @92amartins! Just some adjustments and tests.

File: pupyl/cli.py (hunk 4 above)
Comment (92amartins) [reply pupyl/cli.py:11]: Thanks! I'll have a look.

File: pupyl.py (hunk 1 above)
Comment (policratus) [reply pupyl.py:1]: As we talked, we'll remove the shebang and will create a `shell` script passing parameters to the actual `cli`.

**Comment by policratus** [comment]
@92amartins, I did some changes on the `CLI`, test suites and added the shell script.

**Comment by 92amartins** [comment]
Looks great! I'll test that o/

**Comment by 92amartins** [comment]
@policratus couldn't test the solution in my local environment (possibly because `which python` points to pyenv's shims and stuff).

I think it would be nice to show in README how one can use the CLI, so I made a draft: https://github.com/policratus/pupyl/pull/74/commits/cf3253c7934705cb246f730fd7778eaa304c340d

Feel free to improve on the draft.

**Comment by policratus** [comment]
> @policratus couldn't test the solution in my local environment (possibly because `which python` points to pyenv's shims and stuff).
> 
> I think it would be nice to show in README how one can use the CLI, so I made a draft: [cf3253c](https://github.com/policratus/pupyl/commit/cf3253c7934705cb246f730fd7778eaa304c340d)
> 
> Feel free to improve on the draft.

@92amartins, I made a small change to the `README.md` to describe the mandatory use of `data_dir` and the need to change the CLI script name to `pupyl-cli` to avoid conflicts with the directory containing the module. But I'm not 100% sure if both of these features are the best solution. Let me know what you think.

**Review by 92amartins — COMMENTED** [review]
🚀

**Comment by policratus** [comment]
👍🏼

**Review by policratus — COMMENTED** [review]
👍🏼
